client/src/main/java/com/vesoft/nebula/client/graph/PooledGraphClient.java

```diff
@@ -0,0 +1,152 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+package com.vesoft.nebula.client.graph;
+
+import com.facebook.thrift.TException;
+import com.vesoft.nebula.graph.ErrorCode;
+import com.vesoft.nebula.utils.GraphConstants;
+import java.util.Properties;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Graph Client with thread safe and connection failover able
+ */
+public class PooledGraphClient implements GraphClient {
+    private static final Logger LOG = LoggerFactory.getLogger(PooledGraphClient.class);
+    private final GenericObjectPool<GraphClientImpl> pool;
+    private final ConnectionContext context;
+    private volatile boolean closed = false;
+
+    public PooledGraphClient(Properties props) {
+        this.context = new ConnectionContext()
+                .withHosts(props.getProperty(GraphConstants.CONN_ADDRESS))
+                .withConnectionRetry(props.getProperty(GraphConstants.CONN_RETRY_COUNT, "4"))
```

Comment:
So the connection pool's config is immutable ?

Reply:
currently, only hosts are updatable.

Reply:
It's unnecessary...